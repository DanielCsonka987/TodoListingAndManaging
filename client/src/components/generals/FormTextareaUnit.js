import React, { Component } from 'react'

class FormTextareaUnit extends Component{
    constructor(props){
        super(props)
        this.showBubbleOnScreen = this.showBubbleOnScreen.bind(this)
        this.removeBubbleOnScreen = this.removeBubbleOnScreen.bind(this)
        this.state = {
            bubbleAppearance: false,
        }
    }
    showBubbleOnScreen(){
      this.setState({ bubbleAppearance: true })
    }
    removeBubbleOnScreen(){
      this.setState({ bubbleAppearance: false })
    }
    render(){
        const hiddenTooltipBubble = this.state.bubbleAppearance && this.props.children?
            <span className={'textareaTooltipBubble'}>{this.props.children}</span>  : ''

        return (
            <div className={ 'formTextareaOuterWrapper ' + this.props.classes}>
                <div className='formTextareaInnerWrapper'>
                    <span>
                        <label className='formAndCardLabels' 
                            htmlFor={this.props.id}>{this.props.label}</label>
                        <span className='formTextareaCounter'> (Remained {this.props.maxCharLength-this.props.value.length + ' '}
                            from {this.props.maxCharLength})</span>
                    </span>
                    <textarea className='formTextareaInput'
                        rows={4} maxLength={this.props.maxCharLength}
                        name={this.props.name} id={this.props.id} 
                        value={this.props.value}
                        onChange={this.props.funcChange}  onKeyPress={this.props.funcHitEnter}
                        onFocus={this.showBubbleOnScreen} onMouseOver={this.showBubbleOnScreen}
                        onBlur={this.removeBubbleOnScreen } onMouseOut={this.removeBubbleOnScreen}
                    />
                </div>
                { hiddenTooltipBubble  }
            </div>
            
        )
    }
}

export default FormTextareaUnit;