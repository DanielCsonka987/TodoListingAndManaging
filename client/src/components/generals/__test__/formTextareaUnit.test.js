import React from 'react'
import { waitFor, screen, render } from '@testing-library/react'
import FormTextareaUnit from '../FormTextareaUnit'
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event'
import { toMatchDiffSnapshot } from 'snapshot-diff'
expect.extend({ toMatchDiffSnapshot })

describe('Textarea unit testing', ()=>{
    it('Textarea unit without tooltip, starting with empty input', async ()=>{
        const funcChng = jest.fn()
        const comp = <FormTextareaUnit
            classes='TestingClass'
            id='txt' label='For test:'
            value='' name='txtarea' maxCharLength='80'
            funcChange={funcChng}
        ></FormTextareaUnit>
        const { container, rerender, unmount } = render(comp)
    
        const tree = renderer.create(comp).toJSON()
        expect(tree).toMatchSnapshot()
        
        //revision input existence and its some properties
        const txtr = screen.getByLabelText('For test:')
        expect(txtr).toBeInTheDocument()
        expect(txtr).toHaveClass('formTextareaInput')

        //revise container
        expect(container.firstChild).toHaveClass('TestingClass')
        expect(container.firstChild).toHaveClass('formTextareaOuterWrapper')
    
        //revise inner container
        const inner = container.firstChild.firstChild
        expect(inner).toHaveClass('formTextareaInnerWrapper')

        //revise counter
        const counterTxt = screen.getByText(/^\(Remained/);
        expect(counterTxt).toBeInTheDocument()
        expect(counterTxt).toHaveClass('formTextareaCounter')
        const conunterWholeTxt = document.querySelector('.formTextareaCounter')
        expect(conunterWholeTxt).toHaveTextContent('(Remained 80 from 80)')

        //set focus on input
        userEvent.click(txtr)
        await waitFor(()=>{
            expect(tree).toMatchSnapshot()
        })
        expect(document.querySelector('.textareaTooltipBubble'))
            .not.toBeInTheDocument()

        //typing
        userEvent.type(txtr, 'Stg to test')
        await waitFor(()=>{
            expect(funcChng).toHaveBeenCalledTimes(11)
        })

        //updateing value of component
        const comp2 = <FormTextareaUnit
            classes='TestingClass'
            id='txt' label='For test:'
            value='Stg to test' name='txtarea' maxCharLength='80'
            funcChange={funcChng}
        ></FormTextareaUnit>
        rerender(comp2)

        //revise updated parameters
        const updatedTree = renderer.create(comp2).toJSON()
        expect(updatedTree).toMatchSnapshot()
        expect(document.querySelector('.formTextareaCounter'))
            .toHaveTextContent('(Remained 69 from 80)')
        
        expect(tree).toMatchDiffSnapshot(updatedTree)
        unmount()
    })

    it('Textarea unit with some tooltip', async ()=>{
        const funcChng = jest.fn()
        const comp = <FormTextareaUnit
            id='txt' label='For test:'
            value='' name='txtarea' maxCharLength='80'
            funcChange={funcChng}
        >Testing text here</FormTextareaUnit>
        const { unmount } = render(comp)
    
        const theCompRenderer = renderer.create(comp)
        const tree = theCompRenderer.toJSON()
        expect(tree).toMatchSnapshot()
    
        const txtr = screen.getByLabelText('For test:')

        //set focus on input and test tooltip appearence
        userEvent.click(txtr)
        expect(await screen.findByText('Testing text here')).toBeInTheDocument()
        expect(document.querySelector('.textareaTooltipBubble')).toBeInTheDocument()

        const treeUpdate = theCompRenderer.toJSON()
        expect(tree).toMatchDiffSnapshot(treeUpdate)
    
        unmount()
    })


})