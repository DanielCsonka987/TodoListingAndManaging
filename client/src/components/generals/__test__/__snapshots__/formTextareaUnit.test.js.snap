// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea unit testing Textarea unit with some tooltip 1`] = `
<div
  className="formTextareaOuterWrapper undefined"
>
  <div
    className="formTextareaInnerWrapper"
  >
    <span>
      <label
        className="formAndCardLabels"
        htmlFor="txt"
      >
        For test:
      </label>
      <span
        className="formTextareaCounter"
      >
         (Remained 
        80 
        from 
        80
        )
      </span>
    </span>
    <textarea
      className="formTextareaInput"
      id="txt"
      maxLength="80"
      name="txtarea"
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      rows={4}
      value=""
    />
  </div>
  
</div>
`;

exports[`Textarea unit testing Textarea unit with some tooltip 2`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Textarea unit testing Textarea unit without tooltip, starting with empty input 1`] = `
<div
  className="formTextareaOuterWrapper TestingClass"
>
  <div
    className="formTextareaInnerWrapper"
  >
    <span>
      <label
        className="formAndCardLabels"
        htmlFor="txt"
      >
        For test:
      </label>
      <span
        className="formTextareaCounter"
      >
         (Remained 
        80 
        from 
        80
        )
      </span>
    </span>
    <textarea
      className="formTextareaInput"
      id="txt"
      maxLength="80"
      name="txtarea"
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      rows={4}
      value=""
    />
  </div>
  
</div>
`;

exports[`Textarea unit testing Textarea unit without tooltip, starting with empty input 2`] = `
<div
  className="formTextareaOuterWrapper TestingClass"
>
  <div
    className="formTextareaInnerWrapper"
  >
    <span>
      <label
        className="formAndCardLabels"
        htmlFor="txt"
      >
        For test:
      </label>
      <span
        className="formTextareaCounter"
      >
         (Remained 
        80 
        from 
        80
        )
      </span>
    </span>
    <textarea
      className="formTextareaInput"
      id="txt"
      maxLength="80"
      name="txtarea"
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      rows={4}
      value=""
    />
  </div>
  
</div>
`;

exports[`Textarea unit testing Textarea unit without tooltip, starting with empty input 3`] = `
<div
  className="formTextareaOuterWrapper TestingClass"
>
  <div
    className="formTextareaInnerWrapper"
  >
    <span>
      <label
        className="formAndCardLabels"
        htmlFor="txt"
      >
        For test:
      </label>
      <span
        className="formTextareaCounter"
      >
         (Remained 
        69 
        from 
        80
        )
      </span>
    </span>
    <textarea
      className="formTextareaInput"
      id="txt"
      maxLength="80"
      name="txtarea"
      onBlur={[Function]}
      onChange={
        [MockFunction] {
          "calls": Array [
            Array [
              SyntheticBaseEvent {
                "_reactName": "onChange",
                "_targetInst": null,
                "bubbles": true,
                "cancelable": false,
                "currentTarget": null,
                "defaultPrevented": false,
                "eventPhase": 3,
                "isDefaultPrevented": [Function],
                "isPropagationStopped": [Function],
                "isTrusted": false,
                "nativeEvent": InputEvent {
                  "isTrusted": false,
                },
                "target": <textarea
                  class="formTextareaInput"
                  id="txt"
                  maxlength="80"
                  name="txtarea"
                  rows="4"
                >
                  Stg to test
                </textarea>,
                "timeStamp": 1624874126233,
                "type": "change",
              },
            ],
            Array [
              SyntheticBaseEvent {
                "_reactName": "onChange",
                "_targetInst": null,
                "bubbles": true,
                "cancelable": false,
                "currentTarget": null,
                "defaultPrevented": false,
                "eventPhase": 3,
                "isDefaultPrevented": [Function],
                "isPropagationStopped": [Function],
                "isTrusted": false,
                "nativeEvent": InputEvent {
                  "isTrusted": false,
                },
                "target": <textarea
                  class="formTextareaInput"
                  id="txt"
                  maxlength="80"
                  name="txtarea"
                  rows="4"
                >
                  Stg to test
                </textarea>,
                "timeStamp": 1624874126239,
                "type": "change",
              },
            ],
            Array [
              SyntheticBaseEvent {
                "_reactName": "onChange",
                "_targetInst": null,
                "bubbles": true,
                "cancelable": false,
                "currentTarget": null,
                "defaultPrevented": false,
                "eventPhase": 3,
                "isDefaultPrevented": [Function],
                "isPropagationStopped": [Function],
                "isTrusted": false,
                "nativeEvent": InputEvent {
                  "isTrusted": false,
                },
                "target": <textarea
                  class="formTextareaInput"
                  id="txt"
                  maxlength="80"
                  name="txtarea"
                  rows="4"
                >
                  Stg to test
                </textarea>,
                "timeStamp": 1624874126243,
                "type": "change",
              },
            ],
            Array [
              SyntheticBaseEvent {
                "_reactName": "onChange",
                "_targetInst": null,
                "bubbles": true,
                "cancelable": false,
                "currentTarget": null,
                "defaultPrevented": false,
                "eventPhase": 3,
                "isDefaultPrevented": [Function],
                "isPropagationStopped": [Function],
                "isTrusted": false,
                "nativeEvent": InputEvent {
                  "isTrusted": false,
                },
                "target": <textarea
                  class="formTextareaInput"
                  id="txt"
                  maxlength="80"
                  name="txtarea"
                  rows="4"
                >
                  Stg to test
                </textarea>,
                "timeStamp": 1624874126257,
                "type": "change",
              },
            ],
            Array [
              SyntheticBaseEvent {
                "_reactName": "onChange",
                "_targetInst": null,
                "bubbles": true,
                "cancelable": false,
                "currentTarget": null,
                "defaultPrevented": false,
                "eventPhase": 3,
                "isDefaultPrevented": [Function],
                "isPropagationStopped": [Function],
                "isTrusted": false,
                "nativeEvent": InputEvent {
                  "isTrusted": false,
                },
                "target": <textarea
                  class="formTextareaInput"
                  id="txt"
                  maxlength="80"
                  name="txtarea"
                  rows="4"
                >
                  Stg to test
                </textarea>,
                "timeStamp": 1624874126264,
                "type": "change",
              },
            ],
            Array [
              SyntheticBaseEvent {
                "_reactName": "onChange",
                "_targetInst": null,
                "bubbles": true,
                "cancelable": false,
                "currentTarget": null,
                "defaultPrevented": false,
                "eventPhase": 3,
                "isDefaultPrevented": [Function],
                "isPropagationStopped": [Function],
                "isTrusted": false,
                "nativeEvent": InputEvent {
                  "isTrusted": false,
                },
                "target": <textarea
                  class="formTextareaInput"
                  id="txt"
                  maxlength="80"
                  name="txtarea"
                  rows="4"
                >
                  Stg to test
                </textarea>,
                "timeStamp": 1624874126271,
                "type": "change",
              },
            ],
            Array [
              SyntheticBaseEvent {
                "_reactName": "onChange",
                "_targetInst": null,
                "bubbles": true,
                "cancelable": false,
                "currentTarget": null,
                "defaultPrevented": false,
                "eventPhase": 3,
                "isDefaultPrevented": [Function],
                "isPropagationStopped": [Function],
                "isTrusted": false,
                "nativeEvent": InputEvent {
                  "isTrusted": false,
                },
                "target": <textarea
                  class="formTextareaInput"
                  id="txt"
                  maxlength="80"
                  name="txtarea"
                  rows="4"
                >
                  Stg to test
                </textarea>,
                "timeStamp": 1624874126276,
                "type": "change",
              },
            ],
            Array [
              SyntheticBaseEvent {
                "_reactName": "onChange",
                "_targetInst": null,
                "bubbles": true,
                "cancelable": false,
                "currentTarget": null,
                "defaultPrevented": false,
                "eventPhase": 3,
                "isDefaultPrevented": [Function],
                "isPropagationStopped": [Function],
                "isTrusted": false,
                "nativeEvent": InputEvent {
                  "isTrusted": false,
                },
                "target": <textarea
                  class="formTextareaInput"
                  id="txt"
                  maxlength="80"
                  name="txtarea"
                  rows="4"
                >
                  Stg to test
                </textarea>,
                "timeStamp": 1624874126279,
                "type": "change",
              },
            ],
            Array [
              SyntheticBaseEvent {
                "_reactName": "onChange",
                "_targetInst": null,
                "bubbles": true,
                "cancelable": false,
                "currentTarget": null,
                "defaultPrevented": false,
                "eventPhase": 3,
                "isDefaultPrevented": [Function],
                "isPropagationStopped": [Function],
                "isTrusted": false,
                "nativeEvent": InputEvent {
                  "isTrusted": false,
                },
                "target": <textarea
                  class="formTextareaInput"
                  id="txt"
                  maxlength="80"
                  name="txtarea"
                  rows="4"
                >
                  Stg to test
                </textarea>,
                "timeStamp": 1624874126285,
                "type": "change",
              },
            ],
            Array [
              SyntheticBaseEvent {
                "_reactName": "onChange",
                "_targetInst": null,
                "bubbles": true,
                "cancelable": false,
                "currentTarget": null,
                "defaultPrevented": false,
                "eventPhase": 3,
                "isDefaultPrevented": [Function],
                "isPropagationStopped": [Function],
                "isTrusted": false,
                "nativeEvent": InputEvent {
                  "isTrusted": false,
                },
                "target": <textarea
                  class="formTextareaInput"
                  id="txt"
                  maxlength="80"
                  name="txtarea"
                  rows="4"
                >
                  Stg to test
                </textarea>,
                "timeStamp": 1624874126290,
                "type": "change",
              },
            ],
            Array [
              SyntheticBaseEvent {
                "_reactName": "onChange",
                "_targetInst": null,
                "bubbles": true,
                "cancelable": false,
                "currentTarget": null,
                "defaultPrevented": false,
                "eventPhase": 3,
                "isDefaultPrevented": [Function],
                "isPropagationStopped": [Function],
                "isTrusted": false,
                "nativeEvent": InputEvent {
                  "isTrusted": false,
                },
                "target": <textarea
                  class="formTextareaInput"
                  id="txt"
                  maxlength="80"
                  name="txtarea"
                  rows="4"
                >
                  Stg to test
                </textarea>,
                "timeStamp": 1624874126294,
                "type": "change",
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      rows={4}
      value="Stg to test"
    />
  </div>
  
</div>
`;

exports[`Textarea unit testing Textarea unit without tooltip, starting with empty input 4`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -13,11 +13,11 @@
        </label>
        <span
          className=\\"formTextareaCounter\\"
        >
           (Remained 
-         80 
+         69 
          from 
          80
          )
        </span>
      </span>
@@ -30,10 +30,10 @@
        onChange={[Function mockConstructor]}
        onFocus={[Function bound showBubbleOnScreen]}
        onMouseOut={[Function bound removeBubbleOnScreen]}
        onMouseOver={[Function bound showBubbleOnScreen]}
        rows={4}
-       value=\\"\\"
+       value=\\"Stg to test\\"
      />
    </div>
    
  </div>"
`;
